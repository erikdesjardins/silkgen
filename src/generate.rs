use crate::analyze::{Extents, Nearby, PixelKind};
use crate::sizes::{KicadDim, KicadPos, PixelDim, PixelPos};
use image::{DynamicImage, GenericImageView};
use rand::Rng;
use std::io;
use std::io::Write;
use std::ops::Neg;

pub struct Config {
    pub pixel_pitch: KicadDim,
    pub clearance: KicadDim,
}

pub fn output_file(
    name: &str,
    image: DynamicImage,
    config: Config,
    mut r: impl Rng,
    mut w: impl Write,
) -> Result<(), io::Error> {
    let image = image.into_luma_alpha8();

    log::info!(
        "Total image dimensions: {} x {}",
        image.width(),
        image.height()
    );

    let extents = Extents::from_image(&image);

    log::info!("Extent of significant pixels: {:?}", extents);
    log::info!("Center of significant pixels: {:?}", extents.center());

    sexpr(&mut w, "footprint", |w| {
        w.write_all(b"\"")?;
        w.write_all(name.replace('\"', "").as_bytes())?;
        w.write_all(b"\"\n")?;

        // prelude
        sexpr(w, "version", |w| w.write_all(b"20220630"))?;
        sexpr(w, "generator", |w| w.write_all(b"silkgen"))?;
        sexpr(w, "layer", |w| w.write_all(b"F.Silkscreen"))?;
        sexpr(w, "tedit", |w| w.write_all(b"0"))?;
        sexpr(w, "attr", |w| {
            w.write_all(b"board_only exclude_from_pos_files exclude_from_bom")
        })?;

        // text references
        sexpr(w, "fp_text", |w| {
            w.write_all(b"reference \"G***\" (at 0 0) (layer F.Fab)\n")?;
            w.write_all(b"(effects (font (size 1.524 1.524) (thickness 0.3)))\n")?;
            tstamp(w, &mut r)
        })?;
        sexpr(w, "fp_text", |w| {
            w.write_all(b"value \"LOGO\" (at 0.75 0) (layer F.Fab) hide\n")?;
            w.write_all(b"(effects (font (size 1.524 1.524) (thickness 0.3)))\n")?;
            tstamp(w, &mut r)
        })?;

        // pixels
        for (x, y, pixel) in GenericImageView::pixels(&image) {
            let layers: &[_] = match PixelKind::from_pixel(pixel) {
                PixelKind::Transparent => continue,
                PixelKind::Light => &["F.SilkS"],
                PixelKind::Dark => &["F.Cu", "F.Mask"],
            };
            let nearby = Nearby::from_index(&image, x, y);
            for layer in layers {
                draw_pixel(
                    w,
                    &mut r,
                    PixelPos {
                        x: PixelDim(x),
                        y: PixelDim(y),
                    },
                    &extents,
                    &nearby,
                    &config,
                    layer,
                )?;
            }
        }

        Ok(())
    })
}

fn draw_pixel(
    w: &mut impl Write,
    r: &mut impl Rng,
    top_left: PixelPos,
    extents: &Extents,
    nearby: &Nearby,
    config: &Config,
    layer: &str,
) -> Result<(), io::Error> {
    sexpr(w, "fp_poly", |w| {
        sexpr(w, "pts", |w| {
            // shift by 1 since we base positions on the top left of the pixel
            let center = extents.center() + PixelPos::X1 + PixelPos::Y1;
            // find pixel coords of edges
            let top = top_left.y;
            let bot = top + PixelDim(1);
            let left = top_left.x;
            let right = left + PixelDim(1);
            // find kicad coords of edges
            let kicad_pos = |pos: PixelDim, center_pos: PixelDim| {
                neg_if(
                    pos < center_pos,
                    config.pixel_pitch * pos.abs_diff(center_pos),
                )
            };
            let top = kicad_pos(top, center.y);
            let bot = kicad_pos(bot, center.y);
            let left = kicad_pos(left, center.x);
            let right = kicad_pos(right, center.x);
            // collect points
            let mut points = Vec::new();
            let mut add_points_from =
                |mut x, mut y, horiz, vert, diag, horiz_is_positive, vert_is_positive| {
                    let sub_or_add = |lhs, should_sub, rhs| {
                        if should_sub {
                            lhs - rhs
                        } else {
                            lhs + rhs
                        }
                    };

                    // dark pixels always fill the entire pixel, and don't need special processing
                    if nearby.this == PixelKind::Dark {
                        points.push(KicadPos { x, y });
                        return;
                    }
                    assert_eq!(
                        nearby.this,
                        PixelKind::Light,
                        "should only be light pixels remaining"
                    );

                    // prepare inset dimensions
                    let x_inset = sub_or_add(x, horiz_is_positive, config.clearance);
                    let y_inset = sub_or_add(y, vert_is_positive, config.clearance);

                    // add clearance for directly adjacent dark->light pixel transitions
                    if horiz == PixelKind::Dark {
                        x = x_inset;
                    }
                    if vert == PixelKind::Dark {
                        y = y_inset;
                    }

                    // normal cases: no diagonal inclusion, or already inset on one side or the other
                    if x == x_inset || y == y_inset || diag != PixelKind::Dark {
                        points.push(KicadPos { x, y });
                        return;
                    }

                    // special case: handle diagonal inclusion, splitting the corner into three points
                    //         x  x_inset
                    //
                    // y       +  +---- ...
                    //            |
                    // y_inset +--+
                    //         |
                    //         |
                    //        ...
                    let mut new_points = [
                        KicadPos { x, y: y_inset },
                        KicadPos {
                            x: x_inset,
                            y: y_inset,
                        },
                        KicadPos { x: x_inset, y },
                    ];
                    // handle different coordinate ordering for top right and bottom left
                    if horiz_is_positive != vert_is_positive {
                        new_points.reverse();
                    }
                    points.extend(new_points);
                };
            add_points_from(
                left,
                top,
                nearby.left,
                nearby.top,
                nearby.top_left,
                false,
                false,
            );
            add_points_from(
                right,
                top,
                nearby.right,
                nearby.top,
                nearby.top_right,
                true,
                false,
            );
            add_points_from(
                right,
                bot,
                nearby.right,
                nearby.bot,
                nearby.bot_right,
                true,
                true,
            );
            add_points_from(
                left,
                bot,
                nearby.left,
                nearby.bot,
                nearby.bot_left,
                false,
                true,
            );
            // place points
            for point in points {
                sexpr(w, "xy", |w| write!(w, "{} {}", point.x, point.y))?;
            }
            Ok(())
        })?;
        sexpr(w, "layer", |w| w.write_all(layer.as_bytes()))?;
        sexpr(w, "width", |w| w.write_all(b"0"))?;
        sexpr(w, "fill", |w| w.write_all(b"solid"))?;
        tstamp(w, r)
    })
}

fn tstamp(w: &mut impl Write, r: &mut impl Rng) -> Result<(), io::Error> {
    sexpr(w, "tstamp", |w| {
        let uuid = uuid::Builder::from_random_bytes(r.gen()).into_uuid();
        w.write_all(uuid.to_string().as_bytes())
    })
}

fn sexpr<W: Write, R>(
    w: &mut W,
    name: &str,
    f: impl FnOnce(&mut W) -> Result<R, io::Error>,
) -> Result<R, io::Error> {
    w.write_all(b"(")?;
    w.write_all(name.as_bytes())?;
    w.write_all(b" ")?;
    let r = f(w)?;
    w.write_all(b")\n")?;
    Ok(r)
}

fn neg_if<T: Neg<Output = T>>(cond: bool, x: T) -> T {
    if cond {
        -x
    } else {
        x
    }
}
