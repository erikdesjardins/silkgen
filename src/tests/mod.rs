use crate::generate;
use rand::rngs::StdRng;
use rand::SeedableRng;

fn run_against(file: &[u8]) -> String {
    let image = image::load_from_memory(file).unwrap();

    let config = generate::Config {
        pixel_pitch: "1mm".parse().unwrap(),
        clearance: "0.1mm".parse().unwrap(),
    };

    let rng = StdRng::from_seed(*b"1234567890abcdefghijklmnopqrstuv");

    let mut out = Vec::new();

    generate::output_file("testname", image, config, rng, &mut out).unwrap();

    String::from_utf8(out).unwrap()
}

#[test]
fn basic() {
    insta::assert_snapshot!(run_against(include_bytes!("basic.png")));
}

#[test]
fn clearance() {
    insta::assert_snapshot!(run_against(include_bytes!("clearance.png")));
}

#[test]
fn annoying_dog() {
    insta::assert_snapshot!(run_against(include_bytes!("annoying_dog.png")));
}
