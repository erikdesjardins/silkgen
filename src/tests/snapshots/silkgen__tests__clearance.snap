---
source: src/tests/mod.rs
expression: "run_against(include_bytes!(\"clearance.png\"))"
---
(footprint "testname"
(version 20220630)
(generator silkgen)
(layer F.Silkscreen)
(tedit 0)
(attr board_only exclude_from_pos_files exclude_from_bom)
(fp_text reference "G***" (at 0 0) (layer F.Fab)
(effects (font (size 1.524 1.524) (thickness 0.3)))
(tstamp c45669fa-23d3-4b15-a7c8-d69b555815e1)
)
(fp_text value "LOGO" (at 0.75 0) (layer F.Fab) hide
(effects (font (size 1.524 1.524) (thickness 0.3)))
(tstamp e8692f66-44d9-4022-bcff-21cb4616bb87)
)
(fp_poly (pts (xy -6 -4)
(xy -5 -4)
(xy -5 -3.1)
(xy -6 -3.1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 082a15f5-203b-440c-9ef1-b46049776530)
)
(fp_poly (pts (xy -5 -4)
(xy -4 -4)
(xy -4 -3.1)
(xy -5 -3.1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp df5ea2bf-9422-42e0-b8be-9882e9db0a75)
)
(fp_poly (pts (xy -3 -4)
(xy -2 -4)
(xy -2 -3)
(xy -3 -3)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 40233e12-fc59-4519-9f01-16e978915da5)
)
(fp_poly (pts (xy -3 -4)
(xy -2 -4)
(xy -2 -3)
(xy -3 -3)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp eab1ba90-f6ba-4511-8675-00d9416b041a)
)
(fp_poly (pts (xy -2 -4)
(xy -1 -4)
(xy -1 -3)
(xy -2 -3)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp be8de91f-f686-460f-9f5e-5162fe705b44)
)
(fp_poly (pts (xy -2 -4)
(xy -1 -4)
(xy -1 -3)
(xy -2 -3)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp d33b933a-5654-44bf-bc1b-3deeb6a7a900)
)
(fp_poly (pts (xy 0 -4)
(xy 0.9 -4)
(xy 0.9 -3)
(xy 0 -3)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp e7d0fa16-6761-40eb-9955-03bafab55d6b)
)
(fp_poly (pts (xy 1 -4)
(xy 2 -4)
(xy 2 -3)
(xy 1 -3)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 0261356c-c89c-4e48-8520-96d0252aaab6)
)
(fp_poly (pts (xy 1 -4)
(xy 2 -4)
(xy 2 -3)
(xy 1 -3)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp a5bc418c-2854-4f83-b08e-0bb794c08989)
)
(fp_poly (pts (xy 3 -4)
(xy 4 -4)
(xy 4 -3)
(xy 3 -3)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 6eb4ae6d-6142-41fb-9b25-4cdc21812ee2)
)
(fp_poly (pts (xy 3 -4)
(xy 4 -4)
(xy 4 -3)
(xy 3 -3)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 0e616b94-447a-4ce8-81fe-5cdba5c4c760)
)
(fp_poly (pts (xy 4.1 -4)
(xy 5 -4)
(xy 5 -3)
(xy 4.1 -3)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 2d8800be-3d23-41a7-8cf8-fb04ea68cf9b)
)
(fp_poly (pts (xy -6 -3)
(xy -5 -3)
(xy -5 -2)
(xy -6 -2)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 11411cef-4c7f-4cac-b104-82e47fc4295e)
)
(fp_poly (pts (xy -6 -3)
(xy -5 -3)
(xy -5 -2)
(xy -6 -2)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 797195e2-dbe7-4764-b492-25bfd7e98ed2)
)
(fp_poly (pts (xy -5 -3)
(xy -4 -3)
(xy -4 -2)
(xy -5 -2)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 13d31128-bf74-4c3b-a895-4a5b295fee4d)
)
(fp_poly (pts (xy -5 -3)
(xy -4 -3)
(xy -4 -2)
(xy -5 -2)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp e89bc590-3041-411e-b89d-234aa0ffb59a)
)
(fp_poly (pts (xy -3 -2.9)
(xy -2 -2.9)
(xy -2 -2)
(xy -3 -2)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp f2a704a2-494c-40cd-af42-c4bf7662a5d3)
)
(fp_poly (pts (xy -2 -2.9)
(xy -1 -2.9)
(xy -1 -2)
(xy -2 -2)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp ef4381ff-8cb5-4f28-aaed-0ed61bb810ff)
)
(fp_poly (pts (xy 0 -3)
(xy 0.9 -3)
(xy 0.9 -2)
(xy 0 -2)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 55294132-e072-4a39-a6d2-754faec26042)
)
(fp_poly (pts (xy 1 -3)
(xy 2 -3)
(xy 2 -2)
(xy 1 -2)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 18ace78c-1bcf-4fd5-8f5d-871750fc17d2)
)
(fp_poly (pts (xy 1 -3)
(xy 2 -3)
(xy 2 -2)
(xy 1 -2)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 3437620a-1794-4e25-9306-f877fbe5d2a9)
)
(fp_poly (pts (xy 3 -3)
(xy 4 -3)
(xy 4 -2)
(xy 3 -2)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 48a36b11-75ec-4d6b-8458-583d99c77051)
)
(fp_poly (pts (xy 3 -3)
(xy 4 -3)
(xy 4 -2)
(xy 3 -2)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp b94d0d36-7c7d-41d2-bff1-7c2ac643f371)
)
(fp_poly (pts (xy 4.1 -3)
(xy 5 -3)
(xy 5 -2)
(xy 4.1 -2)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp e65be2ff-cc32-49ef-9ce3-a26ca09870e9)
)
(fp_poly (pts (xy -6 -1)
(xy -5.1 -1)
(xy -5.1 -0.1)
(xy -6 -0.1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 3e6d305c-2609-484b-a63b-cb08d16b6bb8)
)
(fp_poly (pts (xy -5 -1)
(xy -4 -1)
(xy -4 0)
(xy -5 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 4aabb5c1-c4d4-42c5-804a-7980e01619b0)
)
(fp_poly (pts (xy -5 -1)
(xy -4 -1)
(xy -4 0)
(xy -5 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 1fc4e541-b4d1-41aa-b0c9-f0534aeea335)
)
(fp_poly (pts (xy -3 -1)
(xy -2 -1)
(xy -2 0)
(xy -3 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp b20fc505-53b1-4d09-9573-6922c7009fa8)
)
(fp_poly (pts (xy -3 -1)
(xy -2 -1)
(xy -2 0)
(xy -3 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp c47fd5cb-6ff7-4739-8db6-afa305378c67)
)
(fp_poly (pts (xy -1.9 -1)
(xy -1 -1)
(xy -1 -0.1)
(xy -1.9 -0.1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 67ab86c3-e0af-422e-9b93-5847e99ef5a9)
)
(fp_poly (pts (xy 0 -1)
(xy 1 -1)
(xy 1 0)
(xy 0 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 811b6f2d-e8f2-4734-a64e-3a59e9f06f34)
)
(fp_poly (pts (xy 0 -1)
(xy 1 -1)
(xy 1 0)
(xy 0 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 1c34a706-7478-4cee-88a3-d02e1cb2f59e)
)
(fp_poly (pts (xy 1 -1)
(xy 2 -1)
(xy 2 0)
(xy 1 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp b3df1724-992a-4133-9f9f-64ff76bff0d7)
)
(fp_poly (pts (xy 1 -1)
(xy 2 -1)
(xy 2 0)
(xy 1 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 959614ae-4510-4d87-9b19-fe2090ea70a4)
)
(fp_poly (pts (xy 3 -1)
(xy 4 -1)
(xy 4 0)
(xy 3 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp b8032a35-5e7c-4b8a-b8d2-83064eeb8a7c)
)
(fp_poly (pts (xy 3 -1)
(xy 4 -1)
(xy 4 0)
(xy 3 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 1e5df576-a0b4-4e1a-a18a-7742839e7720)
)
(fp_poly (pts (xy 4 -1)
(xy 5 -1)
(xy 5 0)
(xy 4 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 93822236-2664-4b0d-bdf6-3a552d57700e)
)
(fp_poly (pts (xy 4 -1)
(xy 5 -1)
(xy 5 0)
(xy 4 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp bf6fea51-2f18-46f4-bf3b-9e0af4e58bf1)
)
(fp_poly (pts (xy -6 0)
(xy -5 0)
(xy -5 1)
(xy -6 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp ca1e0306-6d97-498f-9395-93f5f7a0e0d9)
)
(fp_poly (pts (xy -6 0)
(xy -5 0)
(xy -5 1)
(xy -6 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 3d8163df-484b-4e0f-8e1d-a38954363550)
)
(fp_poly (pts (xy -5 0)
(xy -4 0)
(xy -4 1)
(xy -5 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp dc13a389-3a8d-418e-a378-29b945b5a6ad)
)
(fp_poly (pts (xy -5 0)
(xy -4 0)
(xy -4 1)
(xy -5 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp c343ff22-daa4-4ae2-a30a-4e06e28224c9)
)
(fp_poly (pts (xy -3 0)
(xy -2 0)
(xy -2 1)
(xy -3 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp e7c107a8-4fe7-425c-ae9b-809041a961c0)
)
(fp_poly (pts (xy -3 0)
(xy -2 0)
(xy -2 1)
(xy -3 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 16903af6-5560-4bf2-b3e8-b7e85a246a54)
)
(fp_poly (pts (xy -2 0)
(xy -1 0)
(xy -1 1)
(xy -2 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp eac16ab0-b639-4ba4-b123-f6facba2d67d)
)
(fp_poly (pts (xy -2 0)
(xy -1 0)
(xy -1 1)
(xy -2 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 1af16dd4-fe77-4612-8c27-ac913412c8df)
)
(fp_poly (pts (xy 0 0)
(xy 1 0)
(xy 1 1)
(xy 0 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 26fc5ea3-dc16-4bf1-86ff-10b9d013e068)
)
(fp_poly (pts (xy 0 0)
(xy 1 0)
(xy 1 1)
(xy 0 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp f1ae58f1-20f6-4bef-b4fe-f3ede92c8a8d)
)
(fp_poly (pts (xy 1.1 0.1)
(xy 2 0.1)
(xy 2 1)
(xy 1.1 1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 722a17d0-b36e-472b-8301-df026e89fa28)
)
(fp_poly (pts (xy 3 0.1)
(xy 3.9 0.1)
(xy 3.9 1)
(xy 3 1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 3f1175dc-79e4-4dc1-96d6-2cef84bc7607)
)
(fp_poly (pts (xy 4 0)
(xy 5 0)
(xy 5 1)
(xy 4 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp de4e646f-1abc-4a06-809c-a8f91e32058e)
)
(fp_poly (pts (xy 4 0)
(xy 5 0)
(xy 5 1)
(xy 4 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp bd6ff71e-562c-45cc-9b8f-19b09c0cdaf9)
)
(fp_poly (pts (xy -6 2)
(xy -5 2)
(xy -5 3)
(xy -6 3)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp bb804fcd-8151-44b0-b7fd-f1a26e33ac97)
)
(fp_poly (pts (xy -6 2)
(xy -5 2)
(xy -5 3)
(xy -6 3)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 7abec0b8-07f3-4b6e-991b-ed95cf95b6d6)
)
(fp_poly (pts (xy -4.9 2)
(xy -4 2)
(xy -4 3)
(xy -4.9 3)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp fc7ad279-7018-4764-951e-826809d38ebf)
)
(fp_poly (pts (xy -3 2)
(xy -2.1 2)
(xy -2.1 3)
(xy -3 3)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp e55fa5fa-bbd2-40e9-8b09-cec51b976826)
)
(fp_poly (pts (xy -2 2)
(xy -1 2)
(xy -1 3)
(xy -2 3)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 7ae678ab-f385-4350-abbb-a60618f2119d)
)
(fp_poly (pts (xy -2 2)
(xy -1 2)
(xy -1 3)
(xy -2 3)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp beac0742-021e-4937-9722-f487597b369c)
)
(fp_poly (pts (xy 0 2)
(xy 1 2)
(xy 1 3)
(xy 0 3)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 28e9fbc2-329d-469d-b053-1b924e4a71b9)
)
(fp_poly (pts (xy 1 2)
(xy 2 2)
(xy 2 2.9)
(xy 1 2.9)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 4aaa03ba-77c3-4604-b8e3-5dfa9ba22d15)
)
(fp_poly (pts (xy 3 2)
(xy 4 2)
(xy 4 2.9)
(xy 3 2.9)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp b15f0fda-5e3c-4a68-87ef-38e7e4a75373)
)
(fp_poly (pts (xy 4 2)
(xy 5 2)
(xy 5 3)
(xy 4 3)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 8eabce04-de5a-4c9c-9dc2-1b30942be3e3)
)
(fp_poly (pts (xy -6 3.1)
(xy -5 3.1)
(xy -5 4)
(xy -6 4)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 68a6fa25-f555-4206-9869-d9c8991aa6f0)
)
(fp_poly (pts (xy -5 3)
(xy -4 3)
(xy -4 4)
(xy -5 4)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 794a1aa3-a3f4-4811-9d68-6575d398235d)
)
(fp_poly (pts (xy -3 3)
(xy -2 3)
(xy -2 4)
(xy -3 4)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 30003a08-3bd5-4f4e-92d7-2c65e1750639)
)
(fp_poly (pts (xy -2 3.1)
(xy -1 3.1)
(xy -1 4)
(xy -2 4)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp d62229e3-a683-4793-a89a-a2c8961d644c)
)
(fp_poly (pts (xy 0 3)
(xy 0.9 3)
(xy 0.9 4)
(xy 0 4)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 480b2136-5f99-43f2-9fc6-0846611e3a33)
)
(fp_poly (pts (xy 1 3)
(xy 2 3)
(xy 2 4)
(xy 1 4)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp acaec272-0bf0-4a06-8dd1-8693f41a510d)
)
(fp_poly (pts (xy 1 3)
(xy 2 3)
(xy 2 4)
(xy 1 4)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 938fceca-3df1-4c17-91a7-44f69d10249c)
)
(fp_poly (pts (xy 3 3)
(xy 4 3)
(xy 4 4)
(xy 3 4)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp 7185dbcc-8eee-41ab-aaf9-e77010d096fe)
)
(fp_poly (pts (xy 3 3)
(xy 4 3)
(xy 4 4)
(xy 3 4)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp dd427eda-9054-4dca-a463-0d24b541a3b8)
)
(fp_poly (pts (xy 4.1 3)
(xy 5 3)
(xy 5 4)
(xy 4.1 4)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp f3d00f23-bf49-4fa5-9cb2-a57ec02a08cb)
)
)

